src/main/java/org/candlepin/subscriptions/product/UpstreamProductData.java

```diff
@@ -145,6 +148,126 @@ public static UpstreamProductData createFromTree(RESTProductTree skuTree) {
+            product.getSku(),
+            attr);
+        return offeringFromUpstream(product.getSku(), productDataSource);
+      }
+    }
+    if (!Objects.equals(umbData.children, existingProductData.children)) {
+      LOGGER.debug(
+          "Must sync SKU={} from data source because child SKUs changed from {} to {}",
+          product.getSku(),
+          existingProductData.children,
+          umbData.children);
+      return offeringFromUpstream(product.getSku(), productDataSource);
+    }
+    if (!Objects.equals(
+        umbData.attrs.get(Attr.DERIVED_SKU), existingProductData.attrs.get(Attr.DERIVED_SKU))) {
+      LOGGER.debug(
+          "Must sync SKU={} from data source because derived SKU changed from {} to {}",
+          product.getSku(),
+          existingProductData.attrs.get(Attr.DERIVED_SKU),
+          umbData.attrs.get(Attr.DERIVED_SKU));
+      return offeringFromUpstream(product.getSku(), productDataSource);
+    }
+    existingProductData.attrs.forEach(umbData::putIfNoConflict);
+    umbData.engOids.addAll(existingProductData.engOids);
+    return Optional.of(umbData.toOffering());
+  }
+
+  private static UpstreamProductData createFromUmbMessage(UmbOperationalProduct product) {
+    UpstreamProductData data = new UpstreamProductData(product.getSku());
+    if (product.getChildSkus() != null) {
+      data.children.addAll(product.getChildSkus());
+    }
+    Arrays.stream(product.getAttributes())
+        .filter(attr -> CODE_TO_ENUM.containsKey(attr.getCode()))
+        .forEach(attr -> data.attrs.put(CODE_TO_ENUM.get(attr.getCode()), attr.getValue()));
+    data.attrs.put(Attr.X_DESCRIPTION, product.getSkuDescription());
+    data.attrs.put(Attr.X_ROLE, product.getRole());
+    return data;
+  }
+
+  private static UpstreamProductData createFromOffering(Offering offering) {
+    UpstreamProductData data = new UpstreamProductData(offering.getSku());
+    data.children.addAll(offering.getChildSkus());
+    data.engOids.addAll(offering.getProductIds());
+    if (StringUtils.hasText(offering.getRole())) {
+      data.attrs.put(Attr.X_ROLE, offering.getRole());
+    }
+    if (StringUtils.hasText(offering.getProductFamily())) {
+      data.attrs.put(Attr.PRODUCT_FAMILY, offering.getProductFamily());
+    }
+    if (StringUtils.hasText(offering.getProductName())) {
+      data.attrs.put(Attr.PRODUCT_NAME, offering.getProductName());
+    }
+    if (StringUtils.hasText(offering.getDescription())) {
+      data.attrs.put(Attr.X_DESCRIPTION, offering.getDescription());
+    }
+    if (StringUtils.hasText(offering.getDerivedSku())) {
+      data.attrs.put(Attr.DERIVED_SKU, offering.getDerivedSku());
+    }
+    if (offering.getSockets() != null) {
+      data.attrs.put(Attr.SOCKET_LIMIT, offering.getSockets().toString());
+    }
+    if (offering.getHypervisorSockets() != null) {
+      data.attrs.put(Attr.SOCKET_LIMIT, offering.getHypervisorSockets().toString());
```

Comment:
Shouldn't the Key be different for each of this. Won't the latter values rewrite the old values here?

Reply:
No, we expect to only ever see either hypervisor cores/sockets or plain cores/sockets. I'll leave a comment in the code as such.